### ISSUE TITLE
Auto-Bed Leveling?

### ISSUE THREAD

**Issue by hayedid** [AIIoT/2.0.x-SapphirePro-BLTouch#2]
Hello.  Thank you for the firmware branch.  I got my BL Touch installed on my Sapphire Pro... and I got the Z-offset set.  However, I am running into few issues:

Essentially, it appears the Z-offset is being ignored when printing.  I can see the value with an M503, and have done an M500 and M501 to ensure the values are saved and active, but the printer appears to print at Z0 rather than at my programmed Z-1.68 Are there some commands I should include as part of my initial g-code to ensure that probing is done before a print?  Have people using this just added a G28 before printing to do an auto bed-level?  Should anything else be added?

Thank you

**Comment by GeraltR** [comment]
Hello!
You try connect Repetier-host and change z-offset. This is last parameter. Check in printer menu actualy value. This value you can change in menu printer but it is'n save.

**Comment by AIIoT** [comment]
M503 command was commenting correctly in the new build.
Please check now to see if the problem is resolved.

**Comment by hayedid** [comment]
Thank you so much!  I will give it a try.

**Comment by hayedid** [comment]
Oh my.  This is really getting frustrating.   I downloaded the latest firmware which is showing as 2.0.4.4. 

Then I followed these steps to come up with a Z offset of -2.38
https://www.youtube.com/watch?v=y_1Kg45APko

This offset is showing on the printer screen.

I am using this startup code.

M107 ; Turn off fan
G90 ; Absolute positioning
M82 ; Extruder in absolute mode
{IF_BED}M190 S{BED}
; Activate all used extruder
{IF_EXT0}M104 T0 S{TEMP0}
G92 E0 ; Reset extruder position
G28 ; Home extruder
G29 P5 ; Auto Bed Level
G1 Z20 F50
; Wait for all used extruders to reach temperature
{IF_EXT0}M109 T0 S{TEMP0}
; Default start code
;G1 Z15 F{Z_TRAVEL_SPEED}

However, when I start printing, Z shows to be at 0.3.  So it is still not taking into account the offset.

Thank you.

**Comment by AIIoT** [comment]
You have not saved the changes to EEPROM to do this execute the command M500 after M851. I tested the commands on Marline 2.0.5.3 work approx.
The simpler method is:

BL Touch calibration of the Z axis offset.
Printer menu:
Configuration> Probe z Offset
save new settings
Configuration> Store settings

My starting G-code is: 
;M900 K0
M107 ;fan off
G28 ;Home X/Y/Z 
G29; FULL BED LEVELING
G1 Z15.0 F300 ;move the head up 15mm
G92 E0 ;zero the extruded length
G1 E3 F100 ;extrude 3mm of feed stock
G92 E0 ;zero the extruded length

**Comment by hayedid** [comment]
Thank you for your reply.  I tried the above.  Something else must be wrong.  I did save the changes with an M500 after M851.  Here are the steps I followed:

Z-Offset Instructions:
1. Home 3D printer
2. M851 Z0 - Reset Z0Offset
3. M500 - Store setting to eeprom
4. M501 - Set active parameters
5. M503 - Display Active Parameters
6. G28 Z - Home Z Axis
7. G1 F60 Z0 - Move nozzle to true 0 offset
8. M211 S0 - Switch off soft endstops
9. Move nozzle towards bed slowly until the paper can barely move
10. Take note of the Z on the printer display (take that number and add the measurement of the calibration sheet or device used)
11. M851 Z X.XX (X.XX being your z offset achieved)
12. M211 S1 - Enable Soft Endstops
13. M500 - Save settings to Eeprom
14. M501 - Set Active Parameters
15. M503 - display current settings

I also tried your method above.  I turned on my printer.  Pressed the blue button (enter), then Configuration, Then probe Z offset (which shows my -2.38), then the blue button (enter), then Store Settings.

I tried your startup code, but it still doesn't go to the negative offset.  I also tried programming it with the bin file in your link (rather than compiling manually). 

Is the configuration file in this code the same one you are using?

Thank you.

**Comment by AIIoT** [comment]
Yes, I using the AIIoT_Config(I'm using sensor 104GT-2 to hotend) bin file from https://github.com/AIIoT/2.0.x-SapphirePro-BLTouch/releases/tag/2.0.5.3 and change offset Z to -2.15
Yes, configuration file in this code is the same i'm using.

**Comment by hayedid** [comment]
hmmm... this is driving me bonkers.  I got the latest 2.0.5.3 code and went through the configuration file.  There was a default value in there for the z offset of 0.3.  I changed that to -2.8.  I then compiled the code and loaded the bin file.  After loading the bin file, I did an Initialize EEPROM and saw my existing value change from -2.81 to -2.8.  This confirmed for me that the firmware I compiled ran, the config was being used, and the the initialize EEPROM was successful.

Then, I tried to print.  The BL Touch checked 9 places, then it proceeded to try printing without accounting for the -2.8 offset.

I'm baffled.
